File: packages/client/src/hybrid.ts
import { ownProp } from "@ktjn/searchable-analysis";
import type {
  DocStoreEntry,
  Manifest,
  VectorShard,
} from "@ktjn/searchable-format";
import { docStoreEntryToHit, fetchDocStoreEntriesByIds } from "./doc-store.js";
import type { ShardCache } from "./fetch-json.js";
import type { Hit, SearchOptions, SearchResult } from "./search.js";
import { resolve } from "./url.js";
import type { VectorHit } from "./vector-search.js";
import {
  bruteForceVectorSearch,
  reciprocalRankFusion,
} from "./vector-search.js";

/**
 * Runs brute-force cosine similarity (`./vector-search.js`) against the
 * manifest's vector shard for `language`, if one was built
 * (docs/guides/vector-search.md#storage-format) -- a language
 * with no vector shard (e.g. an empty partition, or a corpus built
 * without a `vectors` option at all) simply has no vector matches,
 * mirroring how a language with no term shard returns no lexical
 * matches, rather than being an error.
 */
export async function vectorHitsForLanguage(
  manifest: Manifest,
  cache: ShardCache,
  baseUrl: string,
  language: string,
  queryVector: number[],
  limit: number,
): Promise<VectorHit[]> {
  const file = manifest.vectors && ownProp(manifest.vectors.shards, language);
  if (!file) return [];
  const shard = await cache.fetchJson<VectorShard>(resolve(baseUrl, file));
  return bruteForceVectorSearch(queryVector, shard, limit);
}

/** How many candidates each side of a hybrid fusion considers, beyond the caller's own requested `limit` -- RRF/weighted fusion over just the final page size would starve out documents that rank, say, 4th on one side and 4th on the other from ever combining into a top-10 fused result. */
export function fusionCandidateLimit(limit: number): number {
  return Math.max(limit * 3, 30);
}

export async function vectorOnlySearch(
  manifest: Manifest,
  cache: ShardCache,
  baseUrl: string,
  queryVector: number[],
  options: SearchOptions,
): Promise<SearchResult> {
  const language = options.language ?? manifest.defaultLanguage;
  const limit = options.limit ?? 10;
  const vectorHits = await vectorHitsForLanguage(
    manifest,
    cache,
    baseUrl,
    language,
    queryVector,
    limit,
  );
  const docLookup = await fetchDocStoreEntriesByIds(
    manifest,
    cache,
    baseUrl,
    vectorHits.map((v) => v.docId),
  );
  const hits = vectorHits.map((v) =>
    docStoreEntryToHit(v.docId, v.score, docLookup.get(v.docId)),
  );
  return { hits, totalHits: hits.length, language };
}

/** `(v - min) / (max - min)` over `values`, clamped to a constant `1` when every value is identical (nothing to distinguish, and avoids a divide-by-zero) -- used to bring BM25F scores and cosine similarities onto a comparable `[0, 1]` scale before a weighted combination, since only RRF's rank-based default is calibration-free. */
function minMaxNormalize(values: number[]): (value: number) => number {
  if (values.length === 0) return () => 0;
  const min = Math.min(...values);
  const max = Math.max(...values);
  if (max === min) return () => 1;
  return (value: number) => (value - min) / (max - min);
}

/**
 * Combines a lexical result (computed with a wider-than-requested limit
 * so fusion has real candidates to work with, per `fusionCandidateLimit`)
 * with an independent vector search over the same query
 * (docs/guides/vector-search.md#hybrid-search-combining-lexical-and-vector-scores).
 * Pinned hits (docs/guides/pinning.md) are carried over
 * unchanged and excluded from fusion entirely -- a pin is an editorial
 * override, not a candidate for a similarity-based reordering -- so only
 * the *organic* lexical hits are fused with the vector hits, then both
 * are re-merged with the untouched pinned hits in front, truncated to
 * the caller's real `options.limit`.
 */
export async function fuseHybridResult(
  manifest: Manifest,
  cache: ShardCache,
  baseUrl: string,
  lexicalResult: SearchResult,
  vectorHits: VectorHit[],
  options: SearchOptions,
): Promise<SearchResult> {
  const limit = options.limit ?? 10;
  const pinnedHits = lexicalResult.hits.filter((h) => h.pinned);
  const organicLexicalHits = lexicalResult.hits.filter((h) => !h.pinned);
  const pinnedIds = new Set(pinnedHits.map((h) => h.id));

  const nonPinnedVectorHits = vectorHits.filter((v) => !pinnedIds.has(v.docId));

  let combined: Map<number, number>;
  if (options.vectorWeight !== undefined) {
    const vectorWeight = options.vectorWeight;
    const normalizeLexical = minMaxNormalize(
      organicLexicalHits.map((h) => h.score),
    );
    const normalizeVector = minMaxNormalize(
      nonPinnedVectorHits.map((v) => v.score),
    );
    combined = new Map<number, number>();
    for (const hit of organicLexicalHits) {
      combined.set(
        hit.id,
        (combined.get(hit.id) ?? 0) +
          (1 - vectorWeight) * normalizeLexical(hit.score),
      );
    }
    for (const v of nonPinnedVectorHits) {
      combined.set(
        v.docId,
        (combined.get(v.docId) ?? 0) + vectorWeight * normalizeVector(v.score),
      );
    }
  } else {
    combined = reciprocalRankFusion([
      organicLexicalHits.map((h) => h.id),
      nonPinnedVectorHits.map((v) => v.docId),
    ]);
  }

  const ranked = [...combined.entries()].sort((a, b) => b[1] - a[1]);
  const remainingSlots = Math.max(0, limit - pinnedHits.length);
  const topRanked = ranked.slice(0, remainingSlots);

  const lexicalHitById = new Map(organicLexicalHits.map((h) => [h.id, h]));
  const missingIds = topRanked
    .map(([id]) => id)
    .filter((id) => !lexicalHitById.has(id));
  const extraDocLookup = missingIds.length
    ? await fetchDocStoreEntriesByIds(manifest, cache, baseUrl, missingIds)
    : new Map<number, DocStoreEntry>();

  const hits: Hit[] = [
    ...pinnedHits,
    ...topRanked.map(([id, fusedScore]) => {
      const existing = lexicalHitById.get(id);
      return existing
        ? { ...existing, score: fusedScore }
        : docStoreEntryToHit(id, fusedScore, extraDocLookup.get(id));
    }),
  ];

  const totalHits = new Set([...pinnedIds, ...combined.keys()]).size;

  return {
    hits,
    ...(lexicalResult.facets ? { facets: lexicalResult.facets } : {}),
    totalHits,
    language: lexicalResult.language,
    ...(lexicalResult.didYouMean
      ? { didYouMean: lexicalResult.didYouMean }
      : {}),
  };
}

File: python/searchable-client/src/searchable_client/hybrid.py
"""Vector-only and hybrid (lexical+vector fusion) search paths (mirrors
packages/client/src/hybrid.ts).
"""

from dataclasses import replace
from math import isfinite

from searchable_client.doc_store import _fetch_doc_store_entries_by_ids
from searchable_client.errors import (
    InvalidVectorShardError,
    VectorSearchNotConfiguredError,
)
from searchable_client.fetch import ShardCache, resolve_url
from searchable_client.types import (
    Hit,
    Manifest,
    SearchOptions,
    SearchResult,
    vector_shard_from_dict,
)
from searchable_client.vectors import (
    VectorHit,
    brute_force_vector_search,
    reciprocal_rank_fusion,
)


def _load_vector_hits(
    manifest: Manifest,
    cache: ShardCache,
    base_url: str,
    language: str,
    query_vector: list[float],
    limit: int,
) -> list["VectorHit"]:
    # A language without a vector shard (empty partition, or a corpus built
    # without a `vectors` option) simply has no vector matches, mirroring how a
    # language with no term shard returns no lexical matches rather than raising
    # -- same as the TS client (`vectorHitsForLanguage`). Callers that want a
    # loud error (SearchClient.search) check `manifest.vectors` separately.
    vectors = manifest.vectors
    if vectors is None or language not in vectors.shards:
        return []
    try:
        shard = vector_shard_from_dict(
            cache.fetch_json(resolve_url(base_url, vectors.shards[language]))
        )
    except (KeyError, TypeError, ValueError) as exc:
        raise InvalidVectorShardError(
            f"could not parse vector shard for {language!r}: {exc}"
        ) from exc
    if shard.dims != vectors.dims:
        raise InvalidVectorShardError(
            f"vector shard dims {shard.dims} do not match manifest dims {vectors.dims}"
        )
    if shard.quantization != vectors.quantization:
        raise InvalidVectorShardError(
            f"vector shard quantization {shard.quantization!r} does not match manifest "
            f"quantization {vectors.quantization!r}"
        )
    if shard.quantization == "int8" and shard.quant_range is None:
        raise InvalidVectorShardError("int8 vector shard requires quantRange")
    for entry in shard.entries:
        if len(entry.vector) != shard.dims:
            raise InvalidVectorShardError(
                f"vector entry {entry.passage_id!r} has {len(entry.vector)} dims; "
                f"shard requires {shard.dims}"
            )
        if not all(isfinite(value) for value in entry.vector):
            raise InvalidVectorShardError(
                f"vector entry {entry.passage_id!r} contains non-finite values"
            )
    return brute_force_vector_search(query_vector, shard, limit)


def _vector_hit_result(
    vector_hits: list["VectorHit"],
    manifest: Manifest,
    cache: ShardCache,
    base_url: str,
    options: SearchOptions,
    language: str,
) -> SearchResult:
    ids = [hit.doc_id for hit in vector_hits]
    doc_lookup = _fetch_doc_store_entries_by_ids(manifest, cache, base_url, ids)
    hits = [
        Hit(
            id=vector_hit.doc_id,
            score=vector_hit.score,
            url=doc_lookup[vector_hit.doc_id].url if vector_hit.doc_id in doc_lookup else "",
            fields=doc_lookup[vector_hit.doc_id].fields if vector_hit.doc_id in doc_lookup else {},
            external_id=(
                doc_lookup[vector_hit.doc_id].external_id
                if vector_hit.doc_id in doc_lookup
                else None
            ),
            metadata=(
                doc_lookup[vector_hit.doc_id].metadata if vector_hit.doc_id in doc_lookup else None
            ),
            content_hash=(
                doc_lookup[vector_hit.doc_id].content_hash
                if vector_hit.doc_id in doc_lookup
                else None
            ),
        )
        for vector_hit in vector_hits
    ]
    return SearchResult(hits=hits, total_hits=len(vector_hits), language=language)


def _normalize_scores(scores: list[float]) -> dict[float, float]:
    if not scores:
        return {}
    minimum, maximum = min(scores), max(scores)
    if minimum == maximum:
        return {score: 1.0 for score in scores}
    return {score: (score - minimum) / (maximum - minimum) for score in scores}


def _vector_or_hybrid_search(
    query: str,
    manifest: Manifest,
    cache: ShardCache,
    base_url: str,
    options: SearchOptions,
    query_vector: list[float] | None,
    language: str,
) -> SearchResult:
    from searchable_client.search import search

    if query_vector is None:
        raise VectorSearchNotConfiguredError(
            f'Search mode "{options.mode}" requires a query vector'
        )
    limit = max(0, options.limit)
    # RRF/weighted fusion over just the final page size would starve out
    # documents that rank, say, 4th on one side and 4th on the other from ever
    # combining into a top-10 fused result -- same floor as the TS client
    # (fusionCandidateLimit).
    candidate_limit = max(limit * 3, 30)
    vector_hits = _load_vector_hits(
        manifest, cache, base_url, language, query_vector, candidate_limit
    )
    if options.mode == "vector":
        return _vector_hit_result(vector_hits[:limit], manifest, cache, base_url, options, language)

    lexical_options = replace(options, mode="lexical", vector_weight=None, limit=candidate_limit)
    lexical_result = search(query, manifest, cache, base_url, lexical_options)

    # Pinned hits are an editorial override, not candidates for a
    # similarity-based reordering: carry them through unchanged, exclude them
    # (and their vector hits) from fusion entirely, then re-merge in front,
    # truncating to the caller's real limit after -- mirrors the TS client's
    # fuseHybridResult.
    pinned_hits = [hit for hit in lexical_result.hits if hit.pinned]
    organic_lexical = [hit for hit in lexical_result.hits if not hit.pinned]
    pinned_ids = {hit.id for hit in pinned_hits}
    non_pinned_vector_hits = [v for v in vector_hits if v.doc_id not in pinned_ids]
    organic_lexical_by_id = {hit.id: hit for hit in organic_lexical}
    non_pinned_vector_by_id = {v.doc_id: v for v in non_pinned_vector_hits}

    if options.vector_weight is None:
        fused_scores = reciprocal_rank_fusion(
            [[hit.id for hit in organic_lexical], [v.doc_id for v in non_pinned_vector_hits]]
        )
    else:
        if not 0.0 <= options.vector_weight <= 1.0:
            raise ValueError("vector_weight must be between 0 and 1")
        lexical_normalized = _normalize_scores([hit.score for hit in organic_lexical])
        vector_normalized = _normalize_scores([v.score for v in non_pinned_vector_hits])
        fused_scores = {
            doc_id: (1.0 - options.vector_weight)
            * lexical_normalized.get(organic_lexical_by_id[doc_id].score, 0.0)
            + options.vector_weight
            * vector_normalized.get(non_pinned_vector_by_id[doc_id].score, 0.0)
            for doc_id in {*organic_lexical_by_id, *non_pinned_vector_by_id}
        }
    ranked = sorted(fused_scores, key=lambda doc_id: (-fused_scores[doc_id], doc_id))
    remaining_slots = max(0, limit - len(pinned_hits))
    top_ranked = ranked[:remaining_slots]
    missing_ids = [doc_id for doc_id in top_ranked if doc_id not in organic_lexical_by_id]
    doc_lookup = _fetch_doc_store_entries_by_ids(manifest, cache, base_url, missing_ids)
    hits: list[Hit] = list(pinned_hits)
    for doc_id in top_ranked:
        lexical_hit = organic_lexical_by_id.get(doc_id)
        if lexical_hit is not None:
            hits.append(replace(lexical_hit, score=fused_scores[doc_id]))
            continue
        doc = doc_lookup.get(doc_id)
        hits.append(
            Hit(
                id=doc_id,
                score=fused_scores[doc_id],
                url=doc.url if doc else "",
                fields=doc.fields if doc else {},
                external_id=doc.external_id if doc else None,
                metadata=doc.metadata if doc else None,
                content_hash=doc.content_hash if doc else None,
            )
        )
    return SearchResult(
        hits=hits,
        total_hits=len({*pinned_ids, *fused_scores}),
        language=language,
        facets=lexical_result.facets,
        did_you_mean=lexical_result.did_you_mean,
    )
